documentation/puppet-api/v3/all_environment_modules.json

```diff
@@ -0,0 +1,10 @@
+{
+  "$schema":     "http://json-schema.org/draft-04/schema#",
+  "title":       "All Environment Modules",
+  "description": "An array of information about all the modules in all Puppet code environments",
+  "type":        "array",
+  "properties": {
+    "$ref": "#/environment_modules"
+  },
+  "additionalProperties": false
+}
```

Comment:
It isn't clear to me from the spec that "#/environment_modules" is something that a tool would be expected to resolve in an external file.  Following [this example](http://json-schema.org/latest/json-schema-core.html#rfc.section.7), I had been thinking that something more like this might work:

```json
{
  "$schema":     "http://json-schema.org/draft-04/schema#",
  "title":       "All Environment Modules",
  "description": "An array of information about all the modules in all Puppet code environments",
  "type":        "array",
  "definitions": {
    "environment": { "id": "environment_modules.json"}
  },
  "items": {
    "$ref": "#/definitions/environment"
  },
  "additionalProperties": false
}
```
I was hoping to be able to successfully use this with a schema validator tool but have not so far.  I chose to use the Ruby [json-schema library](https://github.com/ruby-json-schema/json-schema).  In my case, I've been running this command:

```
ruby -rjson-schema -e "puts JSON::Validator.validate!('./all_environment_modules.json','response.json')"
```

... where the first arg is the base schema to validate and the second arg is the JSON file to be validated against the schema.  I expect the printed result to be "true" if schema validation succeeds, else an error if validation failed.

For an "all environments" payload from the endpoint, though, the above schema - as well as the one you currently have on the PR - will return "true" even if content on one or more of the array elements is invalid.  I haven't been able to get the tool to appear to successfully navigate into the "environment_modules.json" as part of the validation.

I have, however, been able to create a single schema that appears to properly validate both the single element and array form of the environment_modules payload:

```json
{
  "$schema":     "http://json-schema.org/draft-04/schema#",
  "title":       "Environment Modules",
  "description": "Information about the modules in a Puppet code environment",
  "oneOf": [
    { "$ref": "#/definitions/environment" },
    { "type": "array",
      "items": { "$ref": "#/definitions/environment"}
    }
  ],
  "definitions": {
    "environment": {
      "type": "object",
      "properties": {
        "modules": {
          "description": "The array of modules which exist in an environment.",
          "type": "array",
          "items": {
            "type": "object",
            "properties": {
              "name": {
                "description": "The name of the puppet module",
                "type": "string"
              },
              "version": {
                "description": "The version of the puppet module",
                "type": "string"
              }
            },
            "required": ["name", "version"],
            "additionalProperties": false
          }
        },
        "name": {
          "description": "Name of the environment",
          "type": "string"
        }
      },
      "required": ["modules", "name"],
      "additionalProperties": false
    }
  }
}
```

While this appeared to work fine, the error I get when schema validation fails is pretty generic.  In my case, to force a schema validation failure, I changed the data for one of the module names from a string to a boolean, true.  Here's the error I saw when validating with the "oneOf" schema:

```
.../attribute.rb:18:in `validation_error': The property '#/' of type Array did not match any of the required schemas (JSON::Schema::ValidationError)
```

Another option which worked for me was to just use the original single environment schema, environment_modules.json, but pass an option into the validator indicating that the source JSON has a "list" of the items represented in the schema:

```
ruby -rjson-schema -e "puts JSON::Validator.validate!('./documentation/puppet-api/v3/environment_modules.json','resource.json', :list=>true)"
```

The validation error for the same bad JSON as in my example above was much more useful in this case:

```
attribute.rb:18:in `validation_error': The property '#/0/name' of type TrueClass did not match the following type: string (JSON::Schema::ValidationError)
```

I'm not sure how prevalent a "list" capability is various schema validators.

---

Backing up a level, there doesn't seem to be an option here which is clearly best in my mind.  I'd probably lean at this point toward just leaving the original JSON schema doc, which has just the single environment representation in it, but just add a comment to the [Schema section](https://github.com/puppetlabs/puppetserver/blob/46c32ecfd5ba80a1342e9ed1728677e3d4aff1c7/documentation/puppet-api/v3/environment_modules.markdown#schema) of the "environment_modules.markdown doc which indicates that the linked schema applies to the response for a single environment but that the format would include an array of environments when no query parameter is specified.  Could also give an example of how to do the list form with the Ruby json-schema library, maybe?  I think I'd be good leaving it there and adding to / modifying this later based on any user feedback we might get.  WDYT?